### DISCUSSION TITLE
Port GFF3 I/O methods of scikit-allel to sgkit?

### DISCUSSION THREAD

**Discussion by patrick-koenig** [sgkit-dev/sgkit#652] [Ideas]
Hi sgkit team,

is it planned to port the GFF3 I/O utility methods (especially the gff3_to_dataframe() method) of scikit-allel to sgkit?

**Comment by alimanfoo** [comment]
+1, I use gff3_to_dataframe() often

**Comment by jeromekelleher** [comment]
We don't have anything concrete planned right now @patrick-koenig, but we'd definitely welcome contributions here! It looks to me like a fairly straightforward porting job, so that we return an xarray dataset with defined variables rather than a pandas dataframe?

We'd be happy to help with the details if you'd like to make a pass at it (although folks are on vacation at the moment, so things might be slow)

**Reply by patrick-koenig** [reply]
For an extensive plant genotyping project at my institute I am currently developing a SNP browser for very large SNP matrices, which is based on scikit-allel and zarr in combination with flask in the backend.

A demo instance can be found here: https://divbrowse.ipk-gatersleben.de/shape/ and the git repo is here: https://github.com/IPK-BIT/divbrowse
The dataset in the demo instance consists of 223,387,147 SNPs by 300 genotypes, and larger matrices are in preparation. 

My plan for the future would be to switch from scikit-allel to sgkit in the backend.

So I am strongly interested in having the GFF3 I/O functionality in sgkit as well. :-) I can well imagine contributing the appropriate code to sgkit, but before that I would have to get familiar with sgkit in more detail.

I would suggest importing the GFF/GFF3 files first as pandas.DataFrame and then using xarray.Dataset.from_dataframe to create an xarray.Dataset from it. This way you could easily provide both, a pandas-based and an xarray-based API for GFF/GFF3. 
What do you guys think about this?

Maybe only the pandas.DataFrame is sufficient, because GFF/GFF3 data is rather tabular than array/matrix.

**Reply by jeromekelleher** [reply]
Sounds good to me @patrick-koenig!

**Reply by alimanfoo** [reply]
> Maybe only the pandas.DataFrame is sufficient, because GFF/GFF3 data is rather tabular than array/matrix.

+1 from me

**Reply by alimanfoo** [reply]
Btw, in porting this to sgkit, it might be an opportunity to simplify/improve a little the implementation that's in scikit-allel. The scikit-allel implementation of [gff3_to_dataframe](https://scikit-allel.readthedocs.io/en/stable/_modules/allel/io/gff.html#gff3_to_dataframe) relies on a function called [iter_gff3](https://scikit-allel.readthedocs.io/en/stable/_modules/allel/io/gff.html#iter_gff3) to build an interator of rows. But I did revisit this recently when I was developing a data access package for malariagen, and realised you could just get pandas to parse the GFF3 directly. In the `malariagen_data` package there is an alternative approach, see [read_gff3](https://github.com/malariagen/malariagen-data-python/blob/e1e4ced1a940c90b031ce4068827975bcd750de7/malariagen_data/util.py#L42). Hth.

**Reply by alimanfoo** [reply]
Btw @patrick-koenig cool SNP browser :-)

**Reply by jeromekelleher** [reply]
Yeah, getting pandas to do the parsing directly sounds like a definite win.

**Reply by tomwhite** [reply]
Can GFF/GFF3 files get very large? If so, and we don't want to read whole files into memory, then reading them into an xarray Dataset, (which can be lazy) might be another way to approach this. In the case of small files, you could easily call `to_dataframe()` to convert it to a Pandas dataframe if you wanted to process it with that API.

In terms of implementation, one way would be to use [dask.dataframe.read_csv](https://docs.dask.org/en/latest/generated/dask.dataframe.read_csv.html#dask.dataframe.read_csv), which can read chunks of large files in parallel, before converting it to an xarray Dataset. Unfortunately, there is no `from_dask_dataframe()` for xarray Dataset yet, although there is some work in https://github.com/pydata/xarray/pull/4659 (which appears to have stalled). However, we do already have some code in sgkit for [converting a Dask DataFrame to a dict of Dask arrays](https://github.com/pystatgen/sgkit/blob/718fe3b58b3da2d231bbfa6330a88bf209068c92/sgkit/io/utils.py#L14-L30), so it might not be too hard to use something like this to convert to an xarray Dataset.

**Reply by jeromekelleher** [reply]
@alimanfoo - do we need a ``gff3_to_zarr`` or something, or are the files always small enough to make direct pandas conversion better?

**Comment by alimanfoo** [comment]
It's usually fine just to read from GFF3 to pandas each time in my
experience, takes maybe a couple of seconds to parse a GFF3 with ~10,000
genes (~100,000 features).

**Comment by alimanfoo** [comment]
Biggest GFF3 files I've encountered are maybe 200,000 rows. Not been a
problem reading whole thing into memory as pandas dataframe.

**Reply by tomwhite** [reply]
Thanks @alimanfoo. Perhaps we should just go the direct to Pandas route as originally suggested then...
